File: javascript/src/featurecollection.js
/**
 * @fileoverview Representation of an Earth Engine FeatureCollection.
 */

goog.provide('ee.FeatureCollection');

goog.require('ee.ApiFunction');
goog.require('ee.Collection');
goog.require('ee.ComputedObject');
goog.require('ee.Feature');
goog.require('ee.Geometry');
goog.require('ee.Image');
goog.require('ee.List');
goog.require('ee.Types');
goog.require('ee.arguments');
goog.require('ee.data');
goog.require('goog.array');



/**
 * FeatureCollections can be constructed from the following arguments:
 *   - A string: assumed to be the name of a collection.
 *   - A single geometry.
 *   - A single feature.
 *   - A list of features.
 *   - A GeoJSON FeatureCollection
 *   - A computed object: reinterpreted as a collection.
 *
 * @param {string|number|Array.<*>|ee.ComputedObject|
 *         ee.Geometry|ee.Feature|ee.FeatureCollection} args
 *     The constructor arguments.
 * @param {string=} opt_column The name of the geometry column to use.  Only
 *     useful when working with a named collection.
 * @constructor
 * @extends {ee.Collection}
 * @export
 */
ee.FeatureCollection = function(args, opt_column) {
  // Constructor safety.
  if (!(this instanceof ee.FeatureCollection)) {
    return ee.ComputedObject.construct(ee.FeatureCollection, arguments);
  } else if (args instanceof ee.FeatureCollection) {
    return args;
  }

  if (arguments.length > 2) {
    throw Error(
        'The FeatureCollection constructor takes at most 2 arguments (' +
        arguments.length + ' given)');
  }

  ee.FeatureCollection.initialize();

  // Wrap geometries with features.
  if (args instanceof ee.Geometry) {
    args = new ee.Feature(args);
  }

  // Wrap single features in an array.
  if (args instanceof ee.Feature) {
    args = [args];
  }

  if (ee.Types.isString(args)) {
    // An ID.
    var actualArgs = {'tableId': args};
    if (opt_column) {
      actualArgs['geometryColumn'] = opt_column;
    }
    ee.FeatureCollection.base(
        this, 
        'constructor', 
        new ee.ApiFunction('Collection.loadTable'), actualArgs);
  } else if (Array.isArray(args)) {
    // A list of features.
    ee.FeatureCollection.base(
        this, 'constructor', new ee.ApiFunction('Collection'), {
      'features': goog.array.map(args, function(elem) {
        return new ee.Feature(elem);
      })
    });
  } else if (args instanceof ee.List) {
    // A computed list of features.  This can't get the extra ee.Feature()
    ee.FeatureCollection.base(
        this, 'constructor', new ee.ApiFunction('Collection'),
        {'features': args});
  } else if (args && typeof args === 'object'
      && args['type'] === 'FeatureCollection') {
    ee.FeatureCollection.base(
        this, 'constructor', new ee.ApiFunction('Collection'),
        {'features': args['features'].map(f => new ee.Feature(f))});
  } else if (args instanceof ee.ComputedObject) {
    // A custom object to reinterpret as a FeatureCollection.
    ee.FeatureCollection.base(
        this, 'constructor', args.func, args.args, args.varName);
  } else {
    throw Error('Unrecognized argument type to convert to a ' +
                'FeatureCollection: ' + args);
  }
};
goog.inherits(ee.FeatureCollection, ee.Collection);


/**
 * Whether the class has been initialized with API functions.
 * @type {boolean}
 * @private
 */
ee.FeatureCollection.initialized_ = false;


/**
 * Imports API functions to this class.
 */
ee.FeatureCollection.initialize = function() {
  if (!ee.FeatureCollection.initialized_) {
    ee.ApiFunction.importApi(
        ee.FeatureCollection, 'FeatureCollection', 'FeatureCollection');
    ee.FeatureCollection.initialized_ = true;
  }
};


/**
 * Removes imported API functions from this class.
 */
ee.FeatureCollection.reset = function() {
  ee.ApiFunction.clearApi(ee.FeatureCollection);
  ee.FeatureCollection.initialized_ = false;
};


/**
 * An imperative function that returns a map id and token, suitable for
 * generating a Map overlay.
 *
 * @param {?Object=} opt_visParams The visualization parameters. Currently only
 *     one parameter, 'color', containing an RGB color string is allowed.  If
 *     vis_params isn't specified, then the color #000000 is used.
 * @param {function(!Object, string=)=} opt_callback An async callback.
 *     If not supplied, the call is made synchronously.
 * @return {!ee.data.MapId|undefined} An object which may be passed to
 *     ee.data.getTileUrl or ui.Map.addLayer, including an additional 'image'
 *     field, containing a Collection.draw image wrapping a FeatureCollection
 *     containing this feature. Undefined if a callback was specified.
 * @export
 */
ee.FeatureCollection.prototype.getMap = function(opt_visParams, opt_callback) {
  var args = ee.arguments.extractFromFunction(
      ee.FeatureCollection.prototype.getMap, arguments);

  var painted = /** @type {!ee.Image} */(
      ee.ApiFunction._apply('Collection.draw', {
        'collection': this,
        'color': (args['visParams'] || {})['color'] || '000000'
      }));

  if (args['callback']) {
    painted.getMap(undefined, args['callback']);
  } else {
    return painted.getMap();
  }
};


/**
 * An imperative function that returns all the known information about this
 * collection via an AJAX call.
 *
 * @param {function(ee.data.FeatureCollectionDescription, string=)=}
 *     opt_callback An optional callback. If not supplied, the call is made
 *     synchronously. If supplied, will be called with the first parameter if
 *     successful and the second if unsuccessful.
 * @return {ee.data.FeatureCollectionDescription} A collection description
 *     whose fields include:
 *     - features: a list containing metadata about the features in the
 *           collection.
 *     - properties: an optional dictionary containing the collection's
 *           metadata properties.
 * @export
 */
ee.FeatureCollection.prototype.getInfo = function(opt_callback) {
  return /** @type {ee.data.FeatureCollectionDescription} */(
      ee.FeatureCollection.base(this, 'getInfo', opt_callback));
};


/**
 * Gets a download URL. When the URL is accessed, the FeatureCollection is
 * downloaded in one of several formats.
 * @param {string=} opt_format The format of download, one of:
 *     "csv", "json", "geojson", "kml", "kmz" ("json" outputs GeoJSON). If
 *     unspecified, defaults to "csv".
 * @param {string|!Array<string>=} opt_selectors Feature property names used to
 *     select the attributes to be downloaded. If unspecified, all properties
 *     are included.
 * @param {string=} opt_filename Name of the file to be downloaded; extension is
 *     appended by default. If unspecified, defaults to "table".
 * @param {function(string?, string=)=} opt_callback An optional
 *     callback. If not supplied, the call is made synchronously.
 * @return {string|undefined} A download URL or undefined if a callback was
 *     specified.
 * @export
 */
ee.FeatureCollection.prototype.getDownloadURL = function(
    opt_format, opt_selectors, opt_filename, opt_callback) {
  const args = ee.arguments.extractFromFunction(
      ee.FeatureCollection.prototype.getDownloadURL, arguments);
  const request = {};
  request['table'] = this;
  if (args['format']) {
    request['format'] = args['format'].toUpperCase();
  }
  if (args['filename']) {
    request['filename'] = args['filename'];
  }
  if (args['selectors']) {
    request['selectors'] = args['selectors'];
  }

  if (args['callback']) {
    ee.data.getTableDownloadId(request, function(downloadId, error) {
      if (downloadId) {
        args['callback'](ee.data.makeTableDownloadUrl(downloadId));
      } else {
        args['callback'](null, error);
      }
    });
  } else {
    return ee.data.makeTableDownloadUrl(
        /** @type {!ee.data.DownloadId} */ (
            ee.data.getTableDownloadId(request)));
  }
};


/**
 * Select properties from each Feature in a collection.  It is also
 * possible to call this function with only string arguments; they
 * will be all be interpreted as propertySelectors (varargs).
 *
 * @param {!Array<string>} propertySelectors A list of names or regexes
 *     specifying the attributes to select.
 * @param {!Array<string>=} opt_newProperties A list of new names for the
 *     output properties. Must match the number of properties selected.
 * @param {boolean=} opt_retainGeometry When false, the result will have a
 *     NULL geometry. Defaults to true.
 * @return {!ee.FeatureCollection} The feature collection with selected
 *     properties.
 * @export
 */
ee.FeatureCollection.prototype.select = function(
    propertySelectors, opt_newProperties, opt_retainGeometry) {
  if (ee.Types.isString(propertySelectors)) {
    // Varargs.
    var varargs = Array.prototype.slice.call(arguments);
    return /** @type {!ee.FeatureCollection} */ (this.map(function(feature) {
      return /** @type {!ee.Feature} */(feature).select(varargs);
    }));
  } else {
    // Translate the argument names.
    var args = ee.arguments.extractFromFunction(
        ee.FeatureCollection.prototype.select, arguments);
    return /** @type {!ee.FeatureCollection} */ (this.map(function(feature) {
      return /** @type {!ee.Feature} */(feature).select(args);
    }));
  }
};


/** @override */
ee.FeatureCollection.prototype.name = function() {
  return 'FeatureCollection';
};


/** @override */
ee.FeatureCollection.prototype.elementType = function() {
  return ee.Feature;
};

File: python/ee/featurecollection.py
"""Representation of an Earth Engine FeatureCollection."""

# Using lowercase function naming to match the JavaScript names.
# pylint: disable=g-bad-name

from __future__ import annotations

from typing import Any

from ee import _arg_types
from ee import _utils
from ee import apifunction
from ee import collection
from ee import computedobject
from ee import data
from ee import deprecation
from ee import ee_exception
from ee import ee_list
from ee import ee_types
from ee import feature
from ee import geometry
from ee import image


class FeatureCollection(collection.Collection):
  """A representation of a FeatureCollection."""

  _initialized = False

  # Tell pytype to not complain about dynamic attributes.
  _HAS_DYNAMIC_ATTRIBUTES = True

  @_utils.accept_opt_prefix('opt_column')
  @deprecation.WarnForDeprecatedAsset('args')
  def __init__(
      self,
      args: None | (
              dict[str, Any] |
              list[Any] |
              str |
              feature.Feature |
              geometry.Geometry |
              computedobject.ComputedObject
      ),
      column: Any | None = None,
  ):
    """Constructs a collection features.

    Args:
      args: constructor argument.  One of: 1) A string - assumed to be the name
        of a collection. 2) A geometry. 3) A feature. 4) An array of features.
        5) A GeoJSON FeatureCollection. 6) A computed object - reinterpreted as
        a collection.
      column: The name of the geometry column to use. Only useful with the
        string constructor.

    Raises:
      EEException: if passed something other than the above.
    """
    self.initialize()

    # Wrap geometries with features.
    if isinstance(args, geometry.Geometry):
      args = feature.Feature(args)

    # Wrap single features in an array.
    if isinstance(args, feature.Feature):
      args = [args]

    if ee_types.isString(args):
      # An ID.
      actual_args = {'tableId': args}
      if column:
        actual_args['geometryColumn'] = column
      super().__init__(
          apifunction.ApiFunction.lookup('Collection.loadTable'), actual_args)
    elif isinstance(args, (list, tuple)):
      # A list of features.
      super().__init__(
          apifunction.ApiFunction.lookup('Collection'), {
              'features': [feature.Feature(i) for i in args]
          })
    elif isinstance(args, ee_list.List):
      # A computed list of features.
      super().__init__(
          apifunction.ApiFunction.lookup('Collection'), {
              'features': args
          })
    elif isinstance(args, dict) and args.get('type') == self.name():
      # A GeoJSON FeatureCollection
      super().__init__(
          apifunction.ApiFunction.lookup('Collection'),
          {'features': [feature.Feature(i) for i in args.get('features', [])]})
    elif isinstance(args, computedobject.ComputedObject):
      # A custom object to reinterpret as a FeatureCollection.
      super().__init__(args.func, args.args, args.varName)
    else:
      raise ee_exception.EEException(
          'Unrecognized argument type to convert to a FeatureCollection: %s' %
          args)

  @classmethod
  def initialize(cls) -> None:
    """Imports API functions to this class."""
    if not cls._initialized:
      super().initialize()
      apifunction.ApiFunction.importApi(cls, cls.name(), cls.name())
      cls._initialized = True

  @classmethod
  def reset(cls) -> None:
    """Removes imported API functions from this class."""
    apifunction.ApiFunction.clearApi(cls)
    cls._initialized = False

  def getMapId(
      self, vis_params: dict[str, Any] | None = None
  ) -> dict[str, Any]:
    """Fetch and return a map id and token, suitable for use in a Map overlay.

    Args:
      vis_params: The visualization parameters. Currently only one parameter,
          'color', containing a hex RGB color string is allowed.

    Returns:
      A map ID dictionary as described in ee.data.getMapId, including an
      additional 'image' field containing Collection.draw image wrapping a
      FeatureCollection containing this feature.
    """
    painted = apifunction.ApiFunction.apply_('Collection.draw', {
        'collection': self,
        'color': (vis_params or {}).get('color', '000000')
    })
    return painted.getMapId({})

  def getDownloadURL(
      self,
      filetype: str | None = None,
      selectors: Any | None = None,
      filename: str | None = None,
  ) -> str:
    """Gets a download URL.

    When the URL is accessed, the FeatureCollection is downloaded in one of
    several formats.

    Args:
      filetype: The format of download, one of: "csv", "json", "geojson", "kml",
          "kmz" ("json" outputs GeoJSON). If unspecified, defaults to "csv".
      selectors: Feature property names used to select the attributes to be
          downloaded. If unspecified, all properties are included.
      filename: Name of the file to be downloaded; extension is appended by
          default. If unspecified, defaults to "table".

    Returns:
      A URL to download this FeatureCollection.
    """
    request = {}
    request['table'] = self
    if filetype is not None:
      request['format'] = filetype.upper()
    if filename is not None:
      request['filename'] = filename
    if selectors is not None:
      if isinstance(selectors, (list, tuple)):
        selectors = ','.join(selectors)
      request['selectors'] = selectors
    return data.makeTableDownloadUrl(data.getTableDownloadId(request))

  # Deprecated spelling to match the JS library.
  getDownloadUrl = deprecation.Deprecated('Use getDownloadURL().')(
      getDownloadURL)

  # TODO(user): How to handle type annotations for
  #  feature_collection.select('a', 'b', 'c')?
  # pylint: disable-next=keyword-arg-before-vararg
  def select(
      self,
      propertySelectors: Any,
      newProperties: Any | None = None,
      retainGeometry: bool | str = True,
      *args,
  ) -> FeatureCollection:
    """Select properties from each feature in a collection.

    Args:
      propertySelectors: An array of names or regexes specifying the properties
          to select.
      newProperties: An array of strings specifying the new names for the
          selected properties.  If supplied, the length must match the number
          of properties selected.
      retainGeometry: A boolean.  When false, the result will have no geometry.
      *args: Selector elements as varargs.

    Returns:
      The feature collection with selected properties.
    """
    if len(args) or ee_types.isString(propertySelectors):
      args = list(args)
      if not isinstance(retainGeometry, bool):
        args.insert(0, retainGeometry)
      if newProperties is not None:
        args.insert(0, newProperties)
      args.insert(0, propertySelectors)
      return self.map(lambda feat: feat.select(args, None, True))
    else:
      return self.map(
          # pylint: disable-next=g-long-lambda
          lambda feat: feat.select(
              propertySelectors, newProperties, retainGeometry
          )
      )

  @staticmethod
  def name() -> str:
    return 'FeatureCollection'

  @staticmethod
  def elementType() -> type[feature.Feature]:
    return feature.Feature

  def classify(
      self,
      classifier: _arg_types.Classifier,
      # pylint: disable-next=invalid-name
      outputName: _arg_types.String | None = None,
  ) -> FeatureCollection:
    """Returns the result of classifying each feature in a collection.

    Args:
      classifier: The classifier to use.
      outputName: The name of the output property to be added. This argument is
        ignored if the classifier has more than one output.

    Returns:
      An ee.FeatureCollection.
    """

    return apifunction.ApiFunction.call_(
        self.name() + '.classify', self, classifier, outputName
    )

  def cluster(
      self,
      clusterer: _arg_types.Clusterer,
      # pylint: disable-next=invalid-name
      outputName: _arg_types.String | None = None,
  ) -> FeatureCollection:
    """Returns the results of clustering each feature in a collection.

    Clusters each feature in a collection, adding a new column to each feature
    containing the cluster number to which it has been assigned.

    Args:
      clusterer: The clusterer to use.
      outputName: The name of the output property to be added.
    """

    return apifunction.ApiFunction.call_(
        self.name() + '.cluster', self, clusterer, outputName
    )

  def inverseDistance(
      self,
      range: _arg_types.Number,  # pylint: disable=redefined-builtin
      propertyName: _arg_types.String,  # pylint: disable=invalid-name
      mean: _arg_types.Number,
      stdDev: _arg_types.Number,  # pylint: disable=invalid-name
      gamma: _arg_types.Number | None = None,
      reducer: _arg_types.Reducer | None = None,
  ) -> image.Image:
    """Returns an inverse-distance weighted estimate of the value at each pixel.

    Args:
      range: Size of the interpolation window (in meters).
      propertyName: Name of the numeric property to be estimated.
      mean: Global expected mean.
      stdDev: Global standard deviation.
      gamma: Determines how quickly the estimates tend towards the global mean.
      reducer: Reducer used to collapse the 'propertyName' value of overlapping
        points into a single value.
    """

    return apifunction.ApiFunction.call_(
        self.name() + '.inverseDistance',
        self,
        range,
        propertyName,
        mean,
        stdDev,
        gamma,
        reducer,
    )

  def kriging(
      self,
      propertyName: _arg_types.String,  # pylint: disable=invalid-name
      shape: _arg_types.String,
      range: _arg_types.Number,  # pylint: disable=redefined-builtin
      sill: _arg_types.Number,
      nugget: _arg_types.Number,
      # pylint: disable-next=invalid-name
      maxDistance: _arg_types.Number | None = None,
      reducer: _arg_types.Reducer | None = None,
  ) -> image.Image:
    """Returns the results of sampling a Kriging estimator at each pixel.

    Args:
      propertyName: Property to be estimated (must be numeric).
      shape: Semivariogram shape (one of {exponential, gaussian, spherical}).
      range: Semivariogram range, in meters.
      sill: Semivariogram sill.
      nugget: Semivariogram nugget.
      maxDistance: Radius which determines which features are included in each
        pixel's computation, in meters. Defaults to the semivariogram's range.
      reducer: Reducer used to collapse the 'propertyName' value of overlapping
        points into a single value.
    """

    return apifunction.ApiFunction.call_(
        self.name() + '.kriging',
        self,
        propertyName,
        shape,
        range,
        sill,
        nugget,
        maxDistance,
        reducer,
    )

  @staticmethod
  def loadBigQueryTable(
      table: _arg_types.String,
      geometryColumn: _arg_types.String | None = None,
  ) -> FeatureCollection:
    """Returns a FeatureCollection containing data read from a BigQuery table.

    Args:
      table: Path to BigQuery table in a project.dataset.table format.
      geometryColumn: The name of the column to use as the main feature
        geometry. If not specified, all features will have null geometry.
    """

    return apifunction.ApiFunction.call_(
        'FeatureCollection.loadBigQueryTable',
        table,
        geometryColumn,
    )

  def makeArray(
      self,
      properties: _arg_types.List,
      name: _arg_types.String | None = None,
  ) -> FeatureCollection:
    """Returns a collection with a 1-D Array property for each feature.

    Add a 1-D Array to each feature in a collection by combining a list of
    properties for each feature into a 1-D Array.

    All of the properties must be numeric values. If a feature doesn't contain
    all of the named properties, or any of them aren't numeric, the feature will
    be dropped from the resulting collection.

    Args:
      properties: The properties to select.
      name: The name of the new array property.
    """

    return apifunction.ApiFunction.call_(
        self.name() + '.makeArray', self, properties, name
    )

  @staticmethod
  def randomPoints(
      region: geometry.Geometry,
      points: _arg_types.Integer | None = None,
      seed: _arg_types.Integer | None = None,
      # pylint: disable-next=invalid-name
      maxError: _arg_types.ErrorMargin | None = None,
  ) -> FeatureCollection:
    """Returns a collection of random points.

    Generates points that are uniformly random in the given geometry. If the
    geometry is two-dimensional (polygon or multi-polygon) then the returned
    points are uniformly distributed on the given region of the sphere. If the
    geometry is one-dimensional (linestrings), the returned points are
    interpolated uniformly along the geometry's edges. If the geometry has
    dimension zero (points), the returned points are sampled uniformly from the
    input points. If a multi-geometry of mixed dimension is given, points are
    sampled from the component geometries with the highest dimension.

    Args:
      region: The region to generate points for.
      points: The number of points to generate.
      seed: A seed for the random number generator.
      maxError: The maximum amount of error tolerated when performing any
        necessary reprojection.
    """

    return apifunction.ApiFunction.call_(
        'FeatureCollection.randomPoints', region, points, seed, maxError
    )

  @staticmethod
  def runBigQuery(
      query: _arg_types.String,
      geometryColumn: _arg_types.String | None = None,
      maxBytesBilled: _arg_types.Integer | None = int(1e11),
  ) -> FeatureCollection:
    """Returns a FeatureCollection containing result of a BigQuery query.

    Args:
      query: GoogleSQL query to perform on the BigQuery resources.
      geometryColumn: The name of the column to use as the main feature
        geometry. If not specified, all features will have null geometry.
      maxBytesBilled: Maximum number of bytes billed while processing the query.
        Any BigQuery job that exceeds this limit will fail and won't be billed.
        Defaults to 100GB.
    """

    return apifunction.ApiFunction.call_(
        'FeatureCollection.runBigQuery',
        query,
        geometryColumn,
        maxBytesBilled
    )
